File: snobot_sim/src/main/native/cpp/SnobotSim/SimulatorComponents/navx/SpiNavxSimulator.cpp
/*
 * SpiNavxSimulator.cpp
 *
 *  Created on: Jul 9, 2017
 *      Author: PJ
 */

#include "SnobotSim/SimulatorComponents/navx/SpiNavxSimulator.h"
#include "SnobotSim/Logging/SnobotLogger.h"
#include "MockData/SPIData.h"

SpiNavxSimulator::SpiNavxSimulator(int aPort) :
    NavxSimulator(aPort, 200)
{

}

SpiNavxSimulator::~SpiNavxSimulator()
{

}

void SpiNavxSimulator::HandleWrite()
{

}

void SpiNavxSimulator::HandleTransaction()
{

}

void SpiNavxSimulator::HandleRead()
{
//    uint8_t buffer[199];
//    int count = 0;
//
//    mLastWriteAddress = 0x04;
//    if(mLastWriteAddress == 0x00)
//    {
//        GetWriteConfig(buffer);
//        count = 17 + 1;
//    }
//    else if(mLastWriteAddress == 0x04)
//    {
//        GetCurrentData(buffer, 0x04);
//        count = 86 - 0x04 + 1;
//    }
//    else
//    {
//        SNOBOT_LOG(SnobotLogging::CRITICAL,  "Unknown last write address " << ((int) mLastWriteAddress));
//    }
//
//    buffer[count - 1] = GetCRC(buffer, count - 1);
//
//    HALSIM_SetSPISetValueForRead(mNativePort, buffer, count);
}

uint8_t SpiNavxSimulator::GetCRC(uint8_t* buffer, int length)
{
    int i, j;
    uint8_t crc = 0;

    for (i = 0; i < length; i++)
    {
          crc ^= (int)(0x00ff & buffer[i]);
          for (j = 0; j < 8; j++)
          {
              if ((crc & 0x0001)!=0)
              {
                  crc ^= 0x0091;
              }
              crc >>= 1;
          }
    }
    return crc;
}

File: snobot_sim_java/java/snobot_sim_java/com/snobot/simulator/simulator_components/navx/SpiNavxSimulator.java
package com.snobot.simulator.simulator_components.navx;

import java.nio.ByteBuffer;

import org.apache.log4j.Level;
import org.apache.log4j.Logger;

import com.snobot.simulator.jni.SensorFeedbackJni;
import com.snobot.simulator.simulator_components.ISpiWrapper;

public class SpiNavxSimulator extends NavxSimulator implements ISpiWrapper
{
    private static final Logger sLOGGER = Logger.getLogger(SpiNavxSimulator.class);
    public static final int SPI_NAVX_OFFSET = 200;

    public SpiNavxSimulator(int aSpiPort)
    {
        super(aSpiPort, SPI_NAVX_OFFSET);
    }

    static final int CRC7_POLY = 0x0091;

    private static byte getCRC(byte[] buffer, int length)
    {
        int i, j, crc = 0;

        for (i = 0; i < length; i++)
        {
            crc ^= (int) (0x00ff & buffer[i]);
            for (j = 0; j < 8; j++)
            {
                if ((crc & 0x0001) != 0)
                {
                    crc ^= CRC7_POLY;
                }
                crc >>= 1;
            }
        }
        return (byte) crc;
    }

    @Override
    public void handleRead()
    {

        ByteBuffer lastWriteValue = ByteBuffer.allocateDirect(4);
        SensorFeedbackJni.getSpiLastWrite(mNativePort, lastWriteValue, 4);
        lastWriteValue.rewind();
        int lastWrittenAddress = lastWriteValue.get();

        ByteBuffer withoutCrc = null;

        if (lastWrittenAddress == 0x00)
        {
            withoutCrc = createConfigBuffer();
        }
        else if (lastWrittenAddress == 0x04)
        {
            withoutCrc = createDataBuffer(lastWrittenAddress);
        }
        else
        {
            sLOGGER.log(Level.ERROR, "Unknown last write address " + lastWrittenAddress);
        }

        if (withoutCrc != null)
        {
            byte[] raw_bytes = new byte[withoutCrc.capacity() + 1];
            withoutCrc.get(raw_bytes, 0, withoutCrc.capacity());

            byte crc = getCRC(raw_bytes, withoutCrc.capacity());
            raw_bytes[raw_bytes.length - 1] = crc;

            ByteBuffer toSend = ByteBuffer.allocateDirect(withoutCrc.capacity() + 1);
            toSend.put(raw_bytes);
            toSend.rewind();
            SensorFeedbackJni.setSpiValueForRead(mNativePort, toSend, withoutCrc.capacity() + 1);
        }
    }

    @Override
    public void handleWrite()
    {

    }

    @Override
    public void handleTransaction()
    {

    }
}
